### PR TITLE
Disable the use of pip cache system wide

### PR DESCRIPTION
**Pull request by jvonau**
Disable the use of pip cache system wide, leaves no files hanging around in /root/.pip when disabled.

### CODE REVIEW

**Comment by holta** [comment]
This looks very reasonable:

```
pip config set global.cache-dir false
```

What's the best/related documentation we can link to — to make the ramifications just a bit more clear?

**Comment by jvonau** [comment]
https://stackoverflow.com/questions/9510474/removing-pips-cache
```

(pip maintainer here!)

The specific issue of "installing the wrong version due to caching" issue mentioned in the question was fixed in pip 1.4 (back in 2013!):

    Fix a number of issues related to cleaning up and not reusing build directories. (#413, #709, #634, #602, #939, #865, #948)

Since pip 6.0 (back in 2014!), pip install, pip download and pip wheel commands can be told to avoid using the cache with the --no-cache-dir option. (eg: pip install --no-cache-dir <package>)

Since pip 10.0 (back in 2018!), a pip config command was added, which can be used to configure pip to always ignore the cache -- pip config set global.cache-dir false configures pip to not use the cache "globally" (i.e. in all commands).

Since pip 20.1, pip has a pip cache command to manage the contents of pip's cache.

    pip cache purge removes all the wheel files in the cache.
    pip cache remove matplotlib selectively removes files related to a matplotlib from the cache.

In summary, pip provides a lot of ways to tweak how it uses the cache:

    pip cache remove matplotlib: removes all wheel files related to matplotlib from pip's cache.
    pip cache purge: to clear all wheel files from pip's cache.
    pip config set global.cache-dir false: configure pip to not use the cache "globally" (in all commands)
    pip install --no-cache-dir <package>: install a package without using the cache, for just this run.


```

**Comment by holta** [comment]
I'll fix this with a separate PR:

```
/opt/iiab/iiab/scripts/ansible: line 173: pip: command not found

root@raspberrypi:~# pip3 config set global.cache-dir false
Writing to /root/.config/pip/pip.conf
```

(Above is from Raspberry OS 10 Buster on RPi 4.)

**Comment by holta** [comment]
This yellow error/warning below might be concerning?

When running the latest [scripts/ansible](https://github.com/iiab/iiab/blob/master/scripts/ansible#L173-L179) on Raspberry OS 10 Buster Lite on RPi 4:

> The directory 'false' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple

Refs:

- PR #3020 
- PR #3021

**Comment by jvonau** [comment]
I'll go with aged python3.7 release related.

**Comment by jvonau** [comment]
> root@raspberrypi:~# pip3 config set global.cache-dir false
> Writing to /root/.config/pip/pip.conf
> ```
> 
> (Above is from Raspberry OS 10 Buster on RPi 4.)

https://pip.pypa.io/en/stable/cli/pip_config/
Think /root/.config/pip/pip.conf is the wrong place, better to use /etc/pip.conf by passing --global which results in /etc/pip.conf being modified, that would be a better place for this setting to be system wide.

> This yellow error/warning below might be concerning?
> 
> When running the latest [scripts/ansible](https://github.com/iiab/iiab/blob/master/scripts/ansible#L173-L179) on Raspberry OS 10 Buster Lite on RPi 4:
> 
> > The directory 'false' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
> > Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple

Think that warning is bogus, the warning changes to be whatever the CWD is when using sudo on bullseye so it's not python3.7 related.
```
pi@box:~/pi-gen $ sudo pip3 config list
WARNING: The directory '/home/pi/pi-gen/false' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
global.cache-dir='false'
global.extra-index-url='https://www.piwheels.org/simple'
pi@box:~/pi-gen $ cd ~
pi@box:~ $ sudo pip3 config list
WARNING: The directory '/home/pi/false' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
global.cache-dir='false'
global.extra-index-url='https://www.piwheels.org/simple'

```
Meanwhile a normal user there is no warning
```
pi@box:~ $ pip3 config list
global.cache-dir='false'
global.extra-index-url='https://www.piwheels.org/simple'

```
>Since pip 10.0 (back in 2018!), a pip config command was added, which can be used to configure pip to always ignore the cache -- pip config set global.cache-dir false configures pip to not use the cache "globally" (i.e. in all commands).

Think this is a bit of a lie, all that is happening is the value is set to a nonexistent directory but does turn of the caching, Hide the warning behind /dev/null, file a bug upstream, or just ignore it? Still think --global should be added to the command.

**Comment by holta** [comment]
> > When running the latest [scripts/ansible](https://github.com/iiab/iiab/blob/master/scripts/ansible#L173-L179) on Raspberry OS 10 Buster Lite on RPi 4:
> > > The directory 'false' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
> > > Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
> 
> Think that warning is bogus, the warning changes to be whatever the CWD is when using sudo on bullseye so it's not python3.7 related.

Possibly this might be less serious on RasPiOS Bullseye?

In any case, on RasPiOS Buster, many files appear to be erroneously created when pip caching is turned off:

```
root@raspberrypi:/opt/iiab/iiab# du -hsc false/http/*
24K     false/http/0
68K     false/http/2
76K     false/http/4
88K     false/http/6
68K     false/http/7
60K     false/http/8
192K    false/http/9
88K     false/http/a
1.9M    false/http/b
44K     false/http/c
28K     false/http/d
60K     false/http/f
2.7M    total
```

**Comment by jvonau** [comment]
> > Think that warning is bogus, the warning changes to be whatever the CWD is when using sudo on bullseye so it's not python3.7 related.
> 
> Possibly this might be less serious on RasPiOS Bullseye?
> 

Not present on generated Bullseye images, there is no 'false' directory
```
pi@box:~/pi-gen $ ls -la work/IIAB/stage2iiab/rootfs/opt/iiab/iiab
total 444
drwxr-xr-x  8 root root   4096 Nov  8 17:07 .
drwxr-xr-x  8 root root   4096 Nov  8 17:10 ..
-rw-r--r--  1 root root    444 Nov  8 17:05 ansible.cfg
-rw-r--r--  1 root root     23 Nov  8 17:05 ansible_hosts
-rw-r--r--  1 root root    574 Nov  8 17:05 collections.yml
-rw-r--r--  1 root root    139 Nov  8 17:05 CONTRIBUTING.md
drwxr-xr-x  8 root root   4096 Nov  8 17:05 .git
drwxr-xr-x  2 root root   4096 Nov  8 17:05 .github
-rw-r--r--  1 root root     69 Nov  8 17:05 .gitignore
-rwxr-xr-x  1 root root   1886 Nov  8 17:05 iiab-configure
-rw-r--r--  1 root root    396 Nov  8 17:05 iiab-from-cmdline.yml
-rw-r--r--  1 root root    420 Nov  8 17:05 iiab-from-console.yml
-rwxr-xr-x  1 root root   8984 Nov  8 17:05 iiab-install
-rw-r--r--  1 root root 292300 Nov  8 17:20 iiab-install.log
-rwxr-xr-x  1 root root   3716 Nov  8 17:05 iiab-network
-rw-r--r--  1 root root    239 Nov  8 17:05 iiab-network.yml
-rw-r--r--  1 root root   1388 Nov  8 17:05 iiab-stages.yml
-rw-r--r--  1 root root    227 Nov  8 17:05 install-support.yml
-rw-r--r--  1 root root  18092 Nov  8 17:05 LICENSE
-rw-r--r--  1 root root    914 Nov  8 17:05 LICENSING.md
-rw-r--r--  1 root root   5356 Nov  8 17:05 README.md
drwxr-xr-x 60 root root   4096 Nov  8 17:05 roles
-rw-r--r--  1 root root    251 Nov  8 17:05 run-one-role.yml
-rwxr-xr-x  1 root root   4895 Nov  8 17:05 runrole
-rwxr-xr-x  1 root root    571 Nov  8 17:05 runroles
-rw-r--r--  1 root root    186 Nov  8 17:05 runroles-base.yml
drwxr-xr-x  2 root root   4096 Nov  8 17:05 scripts
drwxr-xr-x  2 root root   4096 Nov  8 17:05 tests
-rw-r--r--  1 root root   1471 Nov  8 17:05 test.yml
-rw-r--r--  1 root root   2404 Nov  8 17:05 .travis.yml
-rw-r--r--  1 root root    220 Nov  8 17:05 unmaintained-roles.txt
drwxr-xr-x  2 root root   4096 Nov  8 17:05 vars
```
> In any case, on RasPiOS Buster, many files appear to be erroneously created when pip caching is turned off:
> 
> ```
> root@raspberrypi:/opt/iiab/iiab# du -hsc false/http/*
> 24K     false/http/0
> 68K     false/http/2
> 76K     false/http/4
> 88K     false/http/6
> 68K     false/http/7
> 60K     false/http/8
> 192K    false/http/9
> 88K     false/http/a
> 1.9M    false/http/b
> 44K     false/http/c
> 28K     false/http/d
> 60K     false/http/f
> 2.7M    total
> ```

Ok it is python3.7 related then, it's actually creating 'false' and using that to store the cache in place of using /root/.pip, file a bug upstream or is now there a better recommended way to turn caching off? Should be a non-issue now that new installs of Buster are not supported given the bogus directory is not being created as above.

**Comment by jvonau** [comment]
found this little nugget  
```
There is a typo in pip config set global.cache-dir false. It should be pip config set global.no-cache-dir false. 
```
with the typo now fixed online at the above url but not in the above summary.

**Comment by holta** [comment]
> found this little nugget
> 
> ```
> There is a typo in pip config set global.cache-dir false. It should be pip config set global.no-cache-dir false. 
> ```
> 
> with the typo now fixed online at the above url but not in the above summary.

Thanks for finding this!

(Should we now make this change to `global.no-cache-dir` ?)

**Comment by holta** [comment]
> (Should we now make this change [from `global.cache-dir`] to `global.no-cache-dir` ?)

@jvonau: With the name of the setting now being reversed, does this mean we now want to set it to `true` instead of `false` ?

**Comment by jvonau** [comment]
As confusing as it is short answer is NO see comment 13 from above url summary below
```
If you like to set the --no-cache-dir option by default, you can put this into pip.conf:

[global]
no-cache-dir = false

Note 1: It's confusing, but to enable the no-cache-dir option you actually have to set it to false. Pretty silly if you ask me... but that's how it is. There is a github issue to fix this.
```
https://github.com/pypa/pip/issues/5011

**Comment by holta** [comment]
What an upstream mess.  Thanks @jvonau for tracking this.

For now I assume we should change...?
`pip3 config --global set global.cache-dir false`

to...
`pip3 config --global set global.no-cache-dir false`
